### ISSUE TITLE
Remove sourcecodegen dependency

### ISSUE THREAD

**Issue by hannosch** [zopefoundation/Zope#71]
For Zope 4 the code from five.pt got integrated directly into Products.PageTemplates.

As part of this a new dependency on sourcecodegen got added. This is only used in `Products.PageTemplates.expression.UntrustedPythonExpr.parse` (https://github.com/zopefoundation/Zope/blob/master/src/Products/PageTemplates/expression.py#L173).

That method also uses `from compiler import parse as ast24_parse`.

I think both of those should no longer be needed, as they are backwards compatibility for old Python versions, so they should be replaced with something more modern. Zope 4 only supports Python 2.7 at this point.

@malthe Do you remember off-hand what it takes to get rid of the old AST code here?

**Comment by malthe** [comment]
I don't recall why this function is being imported as ``ast24_parse`` and not just ``parse`` (its real name), but we can't get rid of the ``sourcecodegen`` module. It generates code!

That is, it prints out Python code given an AST.

**Comment by hannosch** [comment]
Mmh, the description from https://github.com/malthe/sourcecodegen reads: "Python (legacy, 2.4 and below) AST source code generation library".

I thought this was the code used in Python 2.5 and earlier. And with the new AST module in Python 2.6 you used a different approach.

**Comment by stephan-hof** [comment]
I think the root cause is RestrictedPython using the 'compiler' package.
Restricted Python uses the AST generated by the 'compiler' package, which is not compatible with the AST of the 'ast' package. As far as I can see PageTemplates and Chameleon are using the 'ast' package. 

So to use RestrictedPython in PageTemplates the parse method needs to
* create a AST from the 'compiler' package => 'compiler' needs to be imported.
```
        encoded = string.encode('utf-8')
        node = ast24_parse(encoded, 'eval').node
        MutatingWalker.walk(node, self.rm)
```
* converting this compiler.AST into a ast.AST (hope it makes sense to you). To achieve this goal 'sourcecodegen' is used. It converts a compiler.AST back into source code. This source code is then feed into ast.parse to create a ast.AST. You can see this by following `super(UntrustedPythonExpr, self).parse(decoded)`.
```
        string = generate_code(node)
        decoded = string.decode('utf-8')
        value = super(UntrustedPythonExpr, self).parse(decoded)

```

From my point of view this 'AST dances' are not needed anymore when RestrictedPython is ported to python3. In python3 there is only one package to manipulate the AST, called 'ast'.
So the call to `super(UntrustedPythonExpr, self).parse(decoded)` is not needed, which removes the dependency to sourcecodegen.

**Comment by malthe** [comment]
@hannosch I think perhaps you're right. I just took a look at the Chameleon codebase and it has a code generator, too.

**Comment by malthe** [comment]
@stephan-hof good findings. I guess we can revisit once RestrictedPython has been ported?

**Comment by hannosch** [comment]
Ah, that makes sense. @loechel and @icemac have begun porting RestrictedPython over to the new ast module at https://github.com/zopefoundation/RestrictedPython/tree/Python3_update - so this is making progress.

**Comment by sallner** [comment]
Was resolved with #109.
